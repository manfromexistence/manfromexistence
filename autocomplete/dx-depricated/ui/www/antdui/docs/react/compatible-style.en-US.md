---
group:
  title: Advanced
order: 1
title: CSS Compatible
---

Ant Design supports the last 2 versions of modern browsers. If you need to be compatible with legacy browsers, please perform downgrade processing according to actual needs:

## StyleProvider

Please ref [`@ant-design/cssinjs`](https://github.com/ant-design/cssinjs#styleprovider).

## Compatible adjustment

The CSS-in-JS feature of Ant Design uses the ":where" selector by default to lower the CSS selector specificity, reducing the additional cost of adjusting custom styles when upgrading for users. However, the compatibility of the ":where" syntax is relatively poor in older browsers ([compatibility](https://developer.mozilla.org/en-US/docs/Web/CSS/:where#browser_compatibility)). In certain scenarios, if you need to support older browsers (or encounter priority conflicts like TailwindCSS), you can use `@ant-design/cssinjs` to disable the default lowering of specificity (please ensure version consistency with antd).

```tsx
import { StyleProvider } from '@ant-design/cssinjs';

// Config `hashPriority` to `high` instead of default `low`
// Which will remove `:where` wrapper
export default () => (
  <StyleProvider hashPriority="high">
    <MyApp />
  </StyleProvider>
);
```

It will turn `:where` to class selector:

```diff
--  :where(.css-bAMboO).ant-btn {
++  .css-bAMboO.ant-btn {
      color: #fff;
    }
```

Note: After turning off the `:where` downgrade, you may need to manually adjust the priority of some styles. Or you can **use PostCSS plugin** to raise application css selector priority. PostCSS provides many plugins can help on this. e.g:

- [postcss-scopify](https://www.npmjs.com/package/postcss-scopify)
- [postcss-increase-specificity](https://www.npmjs.com/package/postcss-increase-specificity)
- [postcss-add-root-selector](https://www.npmjs.com/package/postcss-add-root-selector)

Raise priority through plugin:

```diff
--  .my-btn {
++  #root .my-btn {
      background: red;
    }
```

## CSS Logical Properties

To unify LTR and RTL styles, Ant Design uses CSS logical properties. For example, the original `margin-left` is replaced by `margin-inline-start`, so that it is the starting position spacing under both LTR and RTL. If you need to be compatible with older browsers, you can configure `transformers` through the `StyleProvider` of `@ant-design/cssinjs`:

```tsx
import { legacyLogicalPropertiesTransformer, StyleProvider } from '@ant-design/cssinjs';

// `transformers` provides a way to transform CSS properties
export default () => (
  <StyleProvider transformers={[legacyLogicalPropertiesTransformer]}>
    <MyApp />
  </StyleProvider>
);
```

When toggled, styles will downgrade CSS logical properties:

```diff
.ant-modal-root {
-- inset: 0;
++ top: 0;
++ right: 0;
++ bottom: 0;
++ left: 0;
}
```

## Rem Adaptation

In responsive web development, there is a need for a convenient and flexible way to achieve page adaptation and responsive design. The `px2remTransformer` transformer can quickly and accurately convert pixel units in style sheets to rem units relative to the root element (HTML tag), enabling the implementation of adaptive and responsive layouts.

```tsx
import { px2remTransformer, StyleProvider } from '@ant-design/cssinjs';

const px2rem = px2remTransformer({
  rootValue: 32, // 32px = 1rem; @default 16
});

export default () => (
  <StyleProvider transformers={[px2rem]}>
    <MyApp />
  </StyleProvider>
);
```

The resulting transformed styles:

```diff
 .px2rem-box {
-  width: 400px;
+  width: 12.5rem;
   background-color: green;
-  font-size: 32px;
+  font-size: 1rem;
   border: 10PX solid #f0f;
 }

 @media only screen and (max-width: 600px) {
   .px2rem-box {
     background-color: red;
-    margin: 10px;
+    margin: 0.3125rem;
   }
 }
```

### Options

<!-- prettier-ignore -->
| Parameter | Description  | Type | Default |
| --- | --- | --- | --- |
| rootValue | Font size of the root element | `number` | 16 |
| precision | Decimal places for the converted value | `number` | 5 |
| mediaQuery | Whether to convert px in media queries | `boolean` | false |

For more details, please refer to: [px2rem.ts#Options](https://github.com/ant-design/cssinjs/blob/master/src/transformers/px2rem.ts)

## Shadow DOM Usage

Since `<style />` tag insertion is different from normal DOM in Shadow DOM scenario, you need to use `StyleProvider` of `@ant-design/cssinjs` to configure the `container` property to set the insertion position:

```tsx
import { StyleProvider } from '@ant-design/cssinjs';
import { createRoot } from 'react-dom/client';

const shadowRoot = someEle.attachShadow({ mode: 'open' });
const container = document.createElement('div');
shadowRoot.appendChild(container);
const root = createRoot(container);

root.render(
  <StyleProvider container={shadowRoot}>
    <MyApp />
  </StyleProvider>,
);
```
