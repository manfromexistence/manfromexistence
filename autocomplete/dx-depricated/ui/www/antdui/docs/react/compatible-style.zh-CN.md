---
group:
  title: 进阶使用
order: 1
title: 样式兼容
---

Ant Design 支持最近 2 个版本的现代浏览器。如果你需要兼容旧版浏览器，请根据实际需求进行降级处理：

## StyleProvider

查看 [`@ant-design/cssinjs`](https://github.com/ant-design/cssinjs#styleprovider).

## `:where` 选择器

Ant Design 的 CSS-in-JS 默认通过 `:where` 选择器降低 CSS Selector 优先级，以减少用户升级时额外调整自定义样式的成本，不过 `:where` 语法的[兼容性](https://developer.mozilla.org/en-US/docs/Web/CSS/:where#browser_compatibility)在低版本浏览器比较差。在某些场景下你如果需要支持旧版浏览器（或与 [TailwindCSS 优先级冲突](https://github.com/ant-design/ant-design/issues/38794#issuecomment-1328262525)），你可以使用 `@ant-design/cssinjs` 取消默认的降权操作（请注意版本保持与 antd 一致）：

```tsx
import { StyleProvider } from '@ant-design/cssinjs';

// `hashPriority` 默认为 `low`，配置为 `high` 后，
// 会移除 `:where` 选择器封装
export default () => (
  <StyleProvider hashPriority="high">
    <MyApp />
  </StyleProvider>
);
```

切换后，样式将从 `:where` 切换为类选择器：

```diff
--  :where(.css-bAMboO).ant-btn {
++  .css-bAMboO.ant-btn {
      color: #fff;
    }
```

注意：关闭 `:where` 降权后，你可能需要手动调整一些样式的优先级。亦或者**使用 PostCSS 插件**提升应用样式的优先级，PostCSS 提供了非常多的插件来调整优先级，你可以自行按需选择，例如：

- [postcss-scopify](https://www.npmjs.com/package/postcss-scopify)
- [postcss-increase-specificity](https://www.npmjs.com/package/postcss-increase-specificity)
- [postcss-add-root-selector](https://www.npmjs.com/package/postcss-add-root-selector)

通过插件配置，将你的 css 样式进行提升：

```diff
--  .my-btn {
++  #root .my-btn {
      background: red;
    }
```

## CSS 逻辑属性

为了统一 LTR 和 RTL 样式，Ant Design 使用了 CSS 逻辑属性。例如原 `margin-left` 使用 `margin-inline-start` 代替，使其在 LTR 和 RTL 下都为起始位置间距。如果你需要兼容旧版浏览器（如 360 浏览器、QQ 浏览器 等等），可以通过 `@ant-design/cssinjs` 的 `StyleProvider` 配置 `transformers` 将其转换：

```tsx
import { legacyLogicalPropertiesTransformer, StyleProvider } from '@ant-design/cssinjs';

// `transformers` 提供预处理功能将样式进行转换
export default () => (
  <StyleProvider transformers={[legacyLogicalPropertiesTransformer]}>
    <MyApp />
  </StyleProvider>
);
```

切换后，样式将降级 CSS 逻辑属性：

```diff
.ant-modal-root {
-- inset: 0;
++ top: 0;
++ right: 0;
++ bottom: 0;
++ left: 0;
}
```

## rem 适配

在响应式网页开发中，需要一种方便且灵活的方式来实现页面的适配和响应式设计。`px2remTransformer` 转换器可以快速而准确地将样式表中的像素单位转换为相对于根元素（HTML 标签）的 rem 单位，实现页面的自适应和响应式布局。

```tsx
import { px2remTransformer, StyleProvider } from '@ant-design/cssinjs';

const px2rem = px2remTransformer({
  rootValue: 32, // 32px = 1rem; @default 16
});

export default () => (
  <StyleProvider transformers={[px2rem]}>
    <MyApp />
  </StyleProvider>
);
```

最终转换后的样式：

```diff
 .px2rem-box {
-  width: 400px;
+  width: 12.5rem;
   background-color: green;
-  font-size: 32px;
+  font-size: 1rem;
   border: 10PX solid #f0f;
 }

 @media only screen and (max-width: 600px) {
   .px2rem-box {
     background-color: red;
-    margin: 10px;
+    margin: 0.3125rem;
   }
 }
```

### 配置项

<!-- prettier-ignore -->
| 参数 | 说明  | 类型 | 默认值 |
| --- | --- | --- | --- |
| rootValue | 根元素字体大小 | `number` | 16 |
| precision | 转换后的小数点位数 | `number` | 5 |
| mediaQuery | 是否转换媒体查询中的 px | `boolean` | false |

详细请参考: [px2rem.ts#Options](https://github.com/ant-design/cssinjs/blob/master/src/transformers/px2rem.ts)

## Shadow DOM 场景

在 Shadow DOM 场景中，由于其添加 `<style />` 标签的方式与普通 DOM 不同，所以需要使用 `@ant-design/cssinjs` 的 `StyleProvider` 配置 `container` 属性用于设置插入位置：

```tsx
import { StyleProvider } from '@ant-design/cssinjs';
import { createRoot } from 'react-dom/client';

const shadowRoot = someEle.attachShadow({ mode: 'open' });
const container = document.createElement('div');
shadowRoot.appendChild(container);
const root = createRoot(container);

root.render(
  <StyleProvider container={shadowRoot}>
    <MyApp />
  </StyleProvider>,
);
```
