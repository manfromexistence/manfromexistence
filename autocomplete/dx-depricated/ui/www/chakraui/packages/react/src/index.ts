export * from "./components"
export * from "./popper"
export { defaultSystem, defaultConfig } from "./preset"
export { defaultBaseConfig } from "./preset-base"
export * from "./styled-system"
