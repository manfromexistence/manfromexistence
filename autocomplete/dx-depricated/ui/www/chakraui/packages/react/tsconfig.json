{
  "extends": "../../tsconfig.json",
  "include": ["src", "__tests__", "__stories__"]
}
