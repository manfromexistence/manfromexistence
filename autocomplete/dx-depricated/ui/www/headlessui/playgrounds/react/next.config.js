/** @type {import('next').NextConfig} */
module.exports = {
  reactStrictMode: false,
  devIndicators: {
    buildActivity: false,
  },
}
