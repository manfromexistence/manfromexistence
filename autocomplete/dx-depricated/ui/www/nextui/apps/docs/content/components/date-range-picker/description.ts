const App = `import {DateRangePicker} from "@nextui-org/react";

export default function App() {
  return (
    <DateRangePicker
      label="Stay duration"
      description="Please enter your stay duration"
      className="max-w-xs"
    />
  );
}`;

const react = {
  "/App.jsx": App,
};

export default {
  ...react,
};
