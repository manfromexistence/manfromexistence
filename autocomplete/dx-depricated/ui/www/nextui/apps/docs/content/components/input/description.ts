const App = `import {Input} from "@nextui-org/react";

export default function App() {
  return (
    <Input
      type="email"
      label="Email"
      defaultValue="junior@nextui.org"
      description="We'll never share your email with anyone else."
      className="max-w-xs"
    />
  );
}`;

const react = {
  "/App.jsx": App,
};

export default {
  ...react,
};
