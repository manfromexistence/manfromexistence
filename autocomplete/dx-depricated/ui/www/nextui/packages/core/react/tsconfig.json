{
  "extends": "../../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "allowJs": true,
    "paths": {
      "tailwind-variants": ["../../../node_modules/tailwind-variants"],
      "@react-aria/interactions": ["../../../node_modules/@react-aria/interactions"]
    }
  },
  "include": ["src", "index.ts"],
  "exclude": ["src/scripts/**", "src/index.ts.backup"]
}
