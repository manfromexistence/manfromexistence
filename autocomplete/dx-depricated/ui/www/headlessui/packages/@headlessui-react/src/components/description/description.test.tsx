import { render } from '@testing-library/react'
import React, { type ReactNode } from 'react'
import { Description, useDescriptions } from './description'

jest.mock('../../hooks/use-id')

it('should be possible to use a DescriptionProvider without using a Description', async () => {
  function Component(props: { children: ReactNode }) {
    let [describedby, DescriptionProvider] = useDescriptions()

    return (
      <DescriptionProvider>
        <div aria-describedby={describedby}>{props.children}</div>
      </DescriptionProvider>
    )
  }

  function Example() {
    return <Component>No description</Component>
  }

  let { container } = render(<Example />)
  expect(container.firstChild).toMatchSnapshot()
})

it('should be possible to use a DescriptionProvider and a single Description, and have them linked', async () => {
  function Component(props: { children: ReactNode }) {
    let [describedby, DescriptionProvider] = useDescriptions()

    return (
      <DescriptionProvider>
        <div aria-describedby={describedby}>{props.children}</div>
      </DescriptionProvider>
    )
  }

  function Example() {
    return (
      <Component>
        <Description>I am a description</Description>
        <span>Contents</span>
      </Component>
    )
  }

  let { container } = render(<Example />)
  expect(container.firstChild).toMatchSnapshot()
})

it('should be possible to use a DescriptionProvider and multiple Description components, and have them linked', async () => {
  function Component(props: { children: ReactNode }) {
    let [describedby, DescriptionProvider] = useDescriptions()

    return (
      <DescriptionProvider>
        <div aria-describedby={describedby}>{props.children}</div>
      </DescriptionProvider>
    )
  }

  function Example() {
    return (
      <Component>
        <Description>I am a description</Description>
        <span>Contents</span>
        <Description>I am also a description</Description>
      </Component>
    )
  }

  let { container } = render(<Example />)
  expect(container.firstChild).toMatchSnapshot()
})
