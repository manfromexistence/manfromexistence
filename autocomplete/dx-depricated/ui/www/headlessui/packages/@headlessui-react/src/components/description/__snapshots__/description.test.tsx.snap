// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be possible to use a DescriptionProvider and a single Description, and have them linked 1`] = `
<div
  aria-describedby="headlessui-description-1"
>
  <p
    data-headlessui-state=""
    id="headlessui-description-1"
  >
    I am a description
  </p>
  <span>
    Contents
  </span>
</div>
`;

exports[`should be possible to use a DescriptionProvider and multiple Description components, and have them linked 1`] = `
<div
  aria-describedby="headlessui-description-1 headlessui-description-2"
>
  <p
    data-headlessui-state=""
    id="headlessui-description-1"
  >
    I am a description
  </p>
  <span>
    Contents
  </span>
  <p
    data-headlessui-state=""
    id="headlessui-description-2"
  >
    I am also a description
  </p>
</div>
`;

exports[`should be possible to use a DescriptionProvider without using a Description 1`] = `
<div>
  No description
</div>
`;
